// clarity-angular/datagrid/datagrid-footer.js
/*
 * Copyright (c) 2016-2017 VMware, Inc. All Rights Reserved.
 * This software is released under MIT license.
 * The full license information can be found in LICENSE in the root directory of this project.
 */
import { Component } from "@angular/core";
import { Selection, SelectionType } from "./providers/selection";
import { HideableColumnService } from "./providers/hideable-column.service";
var DatagridFooter = (function () {
    function DatagridFooter(selection, hideableColumnService) {
        this.selection = selection;
        this.hideableColumnService = hideableColumnService;
        this.activeToggler = false; // When undefined/false we don't get views in the UI toggler component.
        // Causing a bunch of failing tests
        this.listChangeSubscription = [];
        /* reference to the enum so that template can access */
        this.SELECTION_TYPE = SelectionType;
    }
    DatagridFooter.prototype.ngOnInit = function () {
        var _this = this;
        this.listChangeSubscription.push(this.hideableColumnService.columnListChange.subscribe(function (change) {
            var hiddenColumns = change.filter(function (col) { return col; });
            if (hiddenColumns.length > 0) {
                _this.activeToggler = true;
            }
            else {
                _this.activeToggler = false;
            }
        }));
    };
    DatagridFooter.prototype.ngOnDestroy = function () {
        this.listChangeSubscription.forEach(function (sub) {
            sub.unsubscribe();
        });
    };
    return DatagridFooter;
}());
export { DatagridFooter };
DatagridFooter.decorators = [
    { type: Component, args: [{
                selector: "clr-dg-footer",
                template: "\n        <ng-container\n                *ngIf=\"(selection.selectionType === SELECTION_TYPE.Multi) && (selection.current.length > 0)\">\n            <clr-checkbox [clrDisabled]=\"true\" [clrChecked]=\"true\" class=\"datagrid-foot-select\">\n                {{selection.current.length}}\n            </clr-checkbox>\n        </ng-container>\n        <clr-dg-column-toggle *ngIf=\"activeToggler\"></clr-dg-column-toggle>\n        <div class=\"datagrid-foot-description\">\n            <ng-content></ng-content>\n        </div>\n        <ng-content select=\"clr-dg-pagination\"></ng-content>\n    ",
                host: {
                    "[class.datagrid-foot]": "true",
                }
            },] },
];
/** @nocollapse */
DatagridFooter.ctorParameters = function () { return [
    { type: Selection, },
    { type: HideableColumnService, },
]; };
